### PR TITLE
WIP: add support for ipywidgets

### PR DESCRIPTION
**Pull request by jbweston**

### CODE REVIEW

**Comment by jbweston** [comment]
## TODO

+ [x] Add ipywidgets javascript (preferably by embedding it; we maybe don't want to load stuff from the web)
+ [x] Ensure that this works when calling `display()` (only tested when a cell evaluates to a widget)

File: jupyter_sphinx/execute.py [hunk 1]
```diff
@@ -44,6 +46,38 @@ def blank_nb(kernel_name):
     })
 
 
+def extract_widget_state(executor):
+    """Extract ipywidget state from a running ExecutePreprocessor"""
+    # Can only run this function inside 'setup_preprocessor'
+    assert hasattr(executor, 'kc')
+    get_widget = 'import ipywidgets; ipywidgets.Widget.get_manager_state()'
```
Comment (akhmerov) [inline jupyter_sphinx/execute.py]: Instead of executing another cell we should instead send a kernel message

Comment (jbweston) [reply jupyter_sphinx/execute.py]: But this is literally what `run_cell` does. Or is there a specific "give me the widget state" message?

File: jupyter_sphinx/execute.py [hunk 2]
```diff
@@ -308,11 +351,24 @@ def apply(self):
...
                 kernel_name,
                 [nbformat.v4.new_code_cell(node.astext()) for node in nodes],
                 self.config.jupyter_execute_kwargs,
             )
+
+            if widget_state:
+                # Append widget state JSON if any widgets were used in the notebook.
+                # XXX: Can we specify a javascript node directly, rather than a 'raw'
+                #      node of 'html' format?
+                doctree.append(docutils.nodes.raw(
```
Comment (akhmerov) [inline jupyter_sphinx/execute.py]: This can potentially be a large amount of data; should we perhaps save it as a separate file and provide it via `src=...`?

File: jupyter_sphinx/execute.py (hunk 1 above)
Comment (akhmerov) [reply jupyter_sphinx/execute.py]: That's what I remember from @SylvainCorlay's explanation. I just have to dig it out from the source.

Comment (jbweston) [reply jupyter_sphinx/execute.py]: I would be surprised if there was such a message, as I would guess that the concept of widgets does not exist on the kernel level.

Implementing the protocol for executing a cell and getting all the output and display ourselves is quite involved, which is why we just call `run_cell`

Comment (jbweston) [reply jupyter_sphinx/execute.py]: should probably also `drop_defaults=True` in `get_manager_state` to match `embed_widgets.py`

File: jupyter_sphinx/execute.py [hunk 3]
```diff
@@ -308,11 +351,24 @@ def apply(self):
...
+                    text=''.join((
+                        '<script type="application/vnd.jupyter.widget-state+json">',
+                        json.dumps(widget_state),
+                        '</script>')),
+                    format='html'
+                ))
```
Comment (jbweston) [inline jupyter_sphinx/execute.py]: This means that there will be one `<script>` tag with type `application/vnd.jupyter.widget-state+json` *per notebook that contains widgets* in the page. I assume that the widget IDs will be unique, but I don't know how the ipywidgets javascript handles it when there are several state jsons in the page

File: jupyter_sphinx/execute.py (hunk 2 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: We may do. `embed_widgets` does not do this currently.

File: jupyter_sphinx/execute.py [hunk 4]
```diff
@@ -44,6 +46,38 @@ def blank_nb(kernel_name):
...
+    resources = {}
+    if cwd is not None:
+        resources['metadata'] = {'path': cwd}
+    ep = ExecutePreprocessor(**kwargs)
+    with ep.setup_preprocessor(nb, resources, km=km):
+        ep.log.info("Executing notebook with kernel: %s" % ep.kernel_name)
+        nb, resources = super(ExecutePreprocessor, ep).preprocess(nb, resources)
+        info_msg = ep._wait_for_reply(ep.kc.kernel_info())
+        nb.metadata['language_info'] = info_msg['content']['language_info']
+        return extract_widget_state(ep)
```
Comment (jbweston) [inline jupyter_sphinx/execute.py]: We need to be more careful here. Widgets are only a thing for Python kernels

File: jupyter_sphinx/execute.py [hunk 5]
```diff
@@ -195,6 +229,12 @@ def cell_output_to_nodes(cell, data_priority, dir):
                     rawsource=data,
                     language='ipython',
                 ))
+            elif 'javascipt' in mime_type or 'json' in mime_type:
```
Comment (basnijholt) [inline jupyter_sphinx/execute.py]: `javascipt` -> `javascript`

only took me about half an hour to find this bug :joy:

File: jupyter_sphinx/execute.py (hunk 4 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: I suppose the widget protocol could be used from kernels of other languages, but I've never seen this.

We can support only extracting widget state from IPython kernels for the moment, and implement others later if there is demand.

Comment (jbweston) [reply jupyter_sphinx/execute.py]: minimally we would need to only extract if the kernel is a python one.

File: jupyter_sphinx/execute.py (hunk 1 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: we should also do this in a try/catch block in case ipywidgets is not installed in the kernel environment (IIRC not a requirement)

File: jupyter_sphinx/execute.py (hunk 5 above)
Comment (basnijholt) [reply jupyter_sphinx/execute.py]: and it should be 
```python
                if 'json' in mime_type:
                    data = json.dumps(data)
                to_add.append(docutils.nodes.raw(
                    text='<script type="{mime_type}">{data}</script>'
                         .format(mime_type=mime_type, data=data),
```

File: jupyter_sphinx/execute.py [hunk 6]
```diff
@@ -342,6 +398,7 @@ def setup(app):
     app.add_config_value(
         'jupyter_execute_data_priority',
         [
+            'application/vnd.jupyter.widget-view+json',
```
Comment (basnijholt) [inline jupyter_sphinx/execute.py]: Why did you remove `'application/javascript',`?

It is needed to work with javascript outputs.

File: jupyter_sphinx/execute.py (hunk 1 above)
Comment (akhmerov) [reply jupyter_sphinx/execute.py]: It seems you're right, and there isn't a simple kernel client-side API for getting widget state. In a running notebook instance, the widget manager stores all of the widget state, but this is seemingly handled by using gradual communication via comm and state updates.

File: jupyter_sphinx/execute.py (hunk 4 above)
Comment (akhmerov) [inline jupyter_sphinx/execute.py]: `extract_widget_state` will only work properly if:
- The kernel is IPython (or anything running Python more generally, e.g. xonsh).
- ipywidgets are importable

Therefore calling `extract_widget_state` should only be done when it's safe.

In absence of multi-language implementation of the Kernel side of jupyter widgets, I believe a relatively safe heuristic for doing this is to check that the outputs contain at least one widget view.

File: jupyter_sphinx/execute.py (hunk 6 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: but it's not needed for ipywidgets, which is the subject of this merge request

File: jupyter_sphinx/execute.py (hunk 2 above)
Comment (akhmerov) [reply jupyter_sphinx/execute.py]: I think it was designed before heavy-weight widgets like `PlotlyWidget`. Hm, but even the core library has e.g. `ImageWidget`, whose state shouldn't just be embedded into the page.

File: jupyter_sphinx/execute.py (hunk 5 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: in fact, we don't need javascript for ipywidgets to work, so I shoudl remove it from here entirely

Comment (akhmerov) [reply jupyter_sphinx/execute.py]: Let's be more strict with mime type checking and verify for exact equality whenever possible.
So let's only treat `application/javascript` as javascript.

Also, is embedding arbitrary json in a script the right thing to do? It certainly won't work for a plotly figure, there one needs something like [this](https://gitlab.kwant-project.org/solidstate/lectures/blob/master/extra_markdown.tpl#L3-26). It also will unlikely produce an expected result when somebody just outputs JSON without any specialized meaning.

File: jupyter_sphinx/execute.py (hunk 3 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: This is, of course, quite easy to fix: we can simply merge the widget states from each iteration of the loop (taking care to raise an exception in the (hopefully very unlikely) case that 2 widgets in different notebooks have the same ID.

File: jupyter_sphinx/execute.py (hunk 5 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: The second-latest commit made the check an equality check

File: jupyter_sphinx/execute.py (hunk 4 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: What about just wrapping the code to execute in a try/catch block>

Comment (jbweston) [inline jupyter_sphinx/execute.py]: Soon this will be moot: https://github.com/jupyter/nbconvert/pull/900

Comment (jbweston) [reply jupyter_sphinx/execute.py]: And this is targeted for nbconvert 5.5

File: jupyter_sphinx/execute.py (hunk 2 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: Fixed in 766dfb79ea707a6f37b888d95f316796f77330c5

File: jupyter_sphinx/execute.py (hunk 1 above)
Comment (jbweston) [reply jupyter_sphinx/execute.py]: Soon this will be moot: jupyter/nbconvert#900. I think it's fine as a stop-gap solution

**Comment by basnijholt** [comment]
The url of the external json is not correctly parsed I think:

```html
<script type="application/vnd.jupyter.widget-state+json" src="/home/docs/checkouts/readthedocs.org/user_builds/adaptive/checkouts/latest/docs/source/_build/jupyter_execute/tutorial/tutorial.Learner1D_widget-state.json"></script>
```
from https://adaptive.readthedocs.io/en/latest/tutorial/tutorial.Learner1D.html

**Comment by jbweston** [comment]
@akhmerov unfortunately we can't put the widget state in a separate file because of this: https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/html-manager/src/libembed.ts#L36

because they just take the "innerHTML" from the script tag we can't specify it as a separate file

**Comment by SylvainCorlay** [comment]
I merged the main PR to jupyter-sphinx. Would you like to reopen this to master?

**Comment by jbweston** [comment]
Closing to reopen against the upstream repo
